crates/air-lib/air-parser/src/parser/validator.rs

```diff
@@ -383,6 +412,42 @@ impl<'i> VariableValidator<'i> {
     fn met_iterator_definition(&mut self, iterator: &Scalar<'i>, span: Span) {
         self.met_iterator_definitions.insert(iterator.name, span);
     }
+
+    // Meeting xor the validator has to choose one of the two xor branches
+    // and clean up all instructions inserted into other_instr so far.
+    pub fn met_xor(&mut self, xor_span: Span, left_span: Span, right_span: Span) {
```

Comment:
I think that instead of deleting instructions, it's better to think in terms of allowed and prohibited spans. And in xor here you can just add values into an allowed span and then check them in `check_after_next_instr`. With that you can avoid traversing folds and all met instructions.